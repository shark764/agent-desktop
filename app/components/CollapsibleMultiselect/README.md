## Origin
This component was created using Collapsible component as basis.

## Updates

## Usage
CollapsibleMultiselect can receive any HTML elements or React component as it's children and make it selectable so you can store a selected item list in redux.

### ES6
```javascript

import React from 'react';
import CollapsibleMultiselect from 'components/CollapsibleMultiselect';

var App = React.createClass({

  selected: true;

  toggleSelection = () => {
    return !this.selected;
  };

  render: function() {
    return(
      <CollapsibleMultiselect
        title="Single Selector"
        toggleSelection={this.toggleSelection}
        singleToggle={this.selected}
      />
    );
  }

});

export default App;
```

---
## Properties *(Options)*
### **transitionTime** | *number* | default: 400
The number of milliseconds for the open/close transition to take.

### **easing** | *string* | default: 'linear'
The CSS easing method you wish to apply to the open/close transition. This string can be any valid value of CSS `transition-timing-function`. For reference view the [MDN documentation](https://developer.mozilla.org/en/docs/Web/CSS/transition-timing-function).

### **open** | *bool* | default: false
Set to true if you want the CollapsibleMultiselect to begin in the open state. You can also use this prop to manage the state from a parent component.

### **title** | *object*
Defines the title to show in multiselect component.

### **items** | *array*
Array of items desired to be selectable.

### **toggleSelection** | *function* | **required**
Function required to toggle the selection of a child/single item.

### **selectAllBtn** | *bool* | default: undefined
Property that allows a "Select All" button to be displayed (if true, it should receive a "selectAll" function too)

### **selectAll** | *function*
Function that toggles all children items of the component at the same time.

### **singleToggleBtn** | *bool*
Icon that changes it state based on a single toggle selection.

### **selectedItems** | *array*
Array where the component stores the selected items, it is mandatory when the collapsible mode is multiple, since is the array where the component compares the current selected values.

### **toggleShowList** | *function*
Function that stores the visible state of the items in the component.
