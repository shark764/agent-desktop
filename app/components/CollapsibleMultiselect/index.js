/*
 * Copyright Â© 2015-2017 Serenova, LLC. All rights reserved.
 */

/**
 *
 * CollapsibleMultiselect
 *
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';

import Icon from 'components/Icon';
import IconSVG from 'components/IconSVG';

const MainContainer = styled.div`
  padding: 5px;
  display: flex;
  cursor: pointer;
  flex-flow: row wrap;
  &:hover {
    background-color: #def8fe;
  }
`;

const ItemContainer = styled.div`
  display: flex;
  overflow-y: hidden;
  padding-left: 10px;
  padding-top: 5px;
  max-height: 100px;
  cursor: pointer;
  &:hover {
    background-color: #def8fe;
  }
`;

const ItemLabel = styled.label`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 95%;
  cursor: pointer;
`;

const checkStatus = {
  float: 'right',
  marginLeft: 'auto',
};

const checkAllStatus = {
  float: 'right',
  marginLeft: '2%',
};

const unCheckAllStatus = {
  float: 'right',
  marginLeft: '2%',
  opacity: '0.4',
  filter: 'alpha(opacity=40)',
};

const iconClosed = {
  marginTop: '6px',
  height: '8px',
  width: '13.33px',
  marginLeft: 'auto',
  flexShrink: '0',
  fontWeight: '600',
  transition: 'transform 0.5s',
};

const iconOpen = {
  transform: 'rotate(180deg)',
  marginLeft: 'auto',
  height: '8px',
  width: '13.33px',
  flexShrink: '0',
  fontWeight: '600',
};

const caretIconContainer = {
  marginLeft: 'auto',
};

const checkAllContainer = {
  marginLeft: '5px',
};

const iconLoading = {
  float: 'right',
};

class CollapsibleMultiselect extends Component {
  constructor(props) {
    super(props);

    // Defaults the dropdown to be closed
    if (this.props.open) {
      this.state = {
        isClosed: false,
        height: 'auto',
        transition: 'none',
      };
    } else {
      this.state = {
        isClosed: true,
        height: 0,
        transition: `height ${this.props.transitionTime}ms ${
          this.props.easing
        }`,
        overflow: 'hidden',
      };
    }

    //  Setting the component mode (multiselect/single) based on the items list
    if (this.props.items !== undefined) {
      this.collapsibleMode = true;
    } else {
      this.collapsibleMode = false;
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.shouldOpenOnNextCycle) {
      this.continueOpenCollapsible();
    }

    if (
      prevState.height === 'auto' &&
      this.state.shouldSwitchAutoOnNextCycle === true
    ) {
      window.setTimeout(() => {
        // Set small timeout to ensure a true re-render
        this.setState({
          height: 0,
          overflow: 'hidden',
          isClosed: true,
          shouldSwitchAutoOnNextCycle: false,
        });
      }, 50);
    }
  }

  closeCollapsible() {
    this.setState({
      shouldSwitchAutoOnNextCycle: true,
      height: this.inner.offsetHeight,
      transition: `height ${this.props.transitionTime}ms ${this.props.easing}`,
    });
  }

  openCollapsible() {
    this.setState({
      shouldOpenOnNextCycle: true,
    });
  }

  continueOpenCollapsible() {
    this.setState({
      height: this.inner.offsetHeight,
      transition: `height ${this.props.transitionTime}ms ${this.props.easing}`,
      isClosed: false,
      shouldOpenOnNextCycle: false,
    });
  }

  //  Function called when component is multiselect to show/hide items list dropdown
  multiSelectTriggerClick = () => {
    //  Toggles visible state in redux
    if (this.props.toggleShowList !== undefined) {
      this.props.toggleShowList();
    }
    if (this.state.isClosed === true) {
      this.openCollapsible();
    } else {
      this.closeCollapsible();
    }
  };

  handleTransitionEnd = () => {
    // Switch to height auto to make the container responsive
    if (!this.state.isClosed) {
      this.setState({ height: 'auto' });
    } else {
      this.setState({ height: 0 });
    }
  };

  //  Toggles a child item based on an id
  selectItem(selectedItem) {
    this.props.toggleSelection(selectedItem.id);
  }

  //  Toggle all items of the component at the same time based on the ids.
  selectAllItems = (e) => {
    e.stopPropagation();
    const itemsIds = this.props.items.map(
      (collapsibleItem) => collapsibleItem.id
    );
    this.props.selectAll(itemsIds);
  };

  //  Function called when component is single select to toggle a boolean value
  singleSelectTriggerClick = () => {
    this.props.toggleSelection();
  };

  render() {
    const dropdownStyle = {
      height: this.state.height,
      WebkitTransition: this.state.transition,
      msTransition: this.state.transition,
      transition: this.state.transition,
      overflow: this.state.overflow,
      maxHeight: '300px',
      overflowY: 'auto',
    };

    //  We want to render the component just when is not loading and it is multiple list, when it is loading or when it is not multiple (single mode)
    if (
      (!this.props.loading &&
        this.collapsibleMode &&
        this.props.items.length > 0) ||
      this.props.loading ||
      !this.collapsibleMode
    ) {
      return (
        <div>
          <MainContainer
            onClick={
              this.collapsibleMode
                ? this.multiSelectTriggerClick
                : this.singleSelectTriggerClick
            }
          >
            <FormattedMessage {...this.props.title} />
            <div style={checkAllContainer} />
            {this.collapsibleMode &&
              !this.props.loading &&
              `(${this.props.selectedItems.length}/${this.props.items.length})`}
            {this.collapsibleMode &&
              this.props.selectAllBtn && (
              <div style={checkAllContainer} onClick={this.selectAllItems}>
                {this.props.loading ? (
                  <IconSVG
                    id={`${this.props.title}-loading-icon`}
                    name="loading"
                    width="20px"
                    style={iconLoading}
                  />
                ) : (
                  <Icon
                    name="checkStatus"
                    style={
                      this.props.selectedItems.length ===
                        this.props.items.length
                        ? checkAllStatus
                        : unCheckAllStatus
                    }
                  />
                )}
              </div>
            )}

            {this.collapsibleMode &&
              !this.props.loading && (
              <div style={caretIconContainer}>
                <Icon
                  name="caret"
                  style={[iconClosed, !this.state.isClosed && iconOpen]}
                />
              </div>
            )}
            {!this.collapsibleMode &&
              this.props.singleToggleBtn && (
              <div style={checkStatus}>
                <Icon name="checkStatus" style={checkStatus} />
              </div>
            )}
          </MainContainer>
          {this.collapsibleMode && (
            <div
              ref={(outer) => {
                this.outer = outer;
              }}
              style={dropdownStyle}
              onTransitionEnd={this.handleTransitionEnd}
            >
              <div
                ref={(inner) => {
                  this.inner = inner;
                }}
              >
                {this.collapsibleMode &&
                  this.props.items.map((item) => (
                    <ItemContainer
                      key={`${item.id}-title`}
                      title={item.name}
                      onClick={() => this.selectItem(item)}
                      tabIndex="0" // eslint-disable-line
                    >
                      <ItemLabel>
                        {item.name}
                      </ItemLabel>
                      {this.props.selectedItems.includes(item.id) && (
                        <Icon name="checkStatus" style={checkStatus} />
                      )}
                    </ItemContainer>
                  ))}
              </div>
            </div>
          )}
        </div>
      );
    }
    return null;
  }
}

CollapsibleMultiselect.propTypes = {
  transitionTime: PropTypes.number,
  easing: PropTypes.string,
  open: PropTypes.bool,
  title: PropTypes.object,
  items: PropTypes.array,
  toggleSelection: PropTypes.func.isRequired,
  selectAllBtn: PropTypes.bool,
  selectAll: PropTypes.func,
  singleToggleBtn: PropTypes.bool,
  selectedItems: PropTypes.array,
  toggleShowList: PropTypes.func,
  loading: PropTypes.bool,
};

CollapsibleMultiselect.defaultProps = {
  transitionTime: 400,
  easing: 'linear',
};

export default CollapsibleMultiselect;
