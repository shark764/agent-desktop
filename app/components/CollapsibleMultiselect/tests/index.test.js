/*
 * Copyright Â© 2015-2019 Serenova, LLC. All rights reserved.
 */

import React from 'react';
import { shallow } from 'enzyme';
import CollapsibleMultiselect from 'components/CollapsibleMultiselect';

describe('<CollapsibleMultiselect />', () => {
  describe('parent window is the same', () => {
    describe('agent preference menu enabled', () => {
      const title = {
        id: 'abc',
        defaultMessage: 'Single Preference Toggle',
      };
      const rendered = shallow(
        <CollapsibleMultiselect
          title={title}
          toggleSelection={() => {}}
          singleToggleBtn
        />
      );
      it('renders the preference option', () => {
        expect(rendered).toMatchSnapshot();
      });
    });

    const itemList = [
      {
        name: 'item1',
        id: '123',
      },
      {
        name: 'item2',
        id: '321',
      },
    ];
    const selectedItemList = [];
    const title = {
      id: 'abc',
      defaultMessage: 'Multi Preference Toggle',
    };
    describe('multi item preference list rendered', () => {
      describe('preference list rendered', () => {
        const rendered = shallow(
          <CollapsibleMultiselect
            title={title}
            items={itemList}
            toggleSelection={() => {}}
            selectedItems={selectedItemList}
          />
        );
        it('renders the multiple preference options', () => {
          expect(rendered).toMatchSnapshot();
        });
      });
    });
    describe('multi item preference list rendered with select all button', () => {
      describe('preference list rendered', () => {
        const rendered = shallow(
          <CollapsibleMultiselect
            title={title}
            items={itemList}
            toggleSelection={() => {}}
            selectAllBtn
            selectedItems={selectedItemList}
            open
          />
        );
        it('renders the multiple preference options with select all button and open by default', () => {
          expect(rendered).toMatchSnapshot();
        });
      });
    });
  });
});
