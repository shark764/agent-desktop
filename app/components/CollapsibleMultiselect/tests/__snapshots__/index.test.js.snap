// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollapsibleMultiselect /> parent window is the same agent preference menu enabled renders the preference option 1`] = `
<div>
  <styled.div
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Single Preference Toggle"
      id="abc"
      values={Object {}}
    />
    <div
      style={
        Object {
          "marginLeft": "5px",
        }
      }
    />
    <div
      style={
        Object {
          "float": "right",
          "marginLeft": "auto",
        }
      }
    >
      <Icon
        name="checkStatus"
        style={
          Object {
            "float": "right",
            "marginLeft": "auto",
          }
        }
      />
    </div>
  </styled.div>
</div>
`;

exports[`<CollapsibleMultiselect /> parent window is the same multi item preference list rendered preference list rendered renders the multiple preference options 1`] = `
<div>
  <styled.div
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Multi Preference Toggle"
      id="abc"
      values={Object {}}
    />
    <div
      style={
        Object {
          "marginLeft": "5px",
        }
      }
    />
    (0/2)
    <div
      style={
        Object {
          "marginLeft": "auto",
        }
      }
    >
      <Icon
        name="caret"
        style={
          Array [
            Object {
              "flexShrink": "0",
              "fontWeight": "600",
              "height": "8px",
              "marginLeft": "auto",
              "marginTop": "6px",
              "transition": "transform 0.5s",
              "width": "13.33px",
            },
            false,
          ]
        }
      />
    </div>
  </styled.div>
  <div
    onTransitionEnd={[Function]}
    style={
      Object {
        "WebkitTransition": "height 400ms linear",
        "height": 0,
        "maxHeight": "300px",
        "msTransition": "height 400ms linear",
        "overflow": "hidden",
        "overflowY": "auto",
        "transition": "height 400ms linear",
      }
    }
  >
    <div>
      <styled.div
        key="123-title"
        onClick={[Function]}
        tabIndex="0"
        title="item1"
      >
        <styled.label>
          item1
        </styled.label>
      </styled.div>
      <styled.div
        key="321-title"
        onClick={[Function]}
        tabIndex="0"
        title="item2"
      >
        <styled.label>
          item2
        </styled.label>
      </styled.div>
    </div>
  </div>
</div>
`;

exports[`<CollapsibleMultiselect /> parent window is the same multi item preference list rendered with select all button preference list rendered renders the multiple preference options with select all button and open by default 1`] = `
<div>
  <styled.div
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Multi Preference Toggle"
      id="abc"
      values={Object {}}
    />
    <div
      style={
        Object {
          "marginLeft": "5px",
        }
      }
    />
    (0/2)
    <div
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "5px",
        }
      }
    >
      <Icon
        name="checkStatus"
        style={
          Object {
            "filter": "alpha(opacity=40)",
            "float": "right",
            "marginLeft": "2%",
            "opacity": "0.4",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "marginLeft": "auto",
        }
      }
    >
      <Icon
        name="caret"
        style={
          Array [
            Object {
              "flexShrink": "0",
              "fontWeight": "600",
              "height": "8px",
              "marginLeft": "auto",
              "marginTop": "6px",
              "transition": "transform 0.5s",
              "width": "13.33px",
            },
            Object {
              "flexShrink": "0",
              "fontWeight": "600",
              "height": "8px",
              "marginLeft": "auto",
              "transform": "rotate(180deg)",
              "width": "13.33px",
            },
          ]
        }
      />
    </div>
  </styled.div>
  <div
    onTransitionEnd={[Function]}
    style={
      Object {
        "WebkitTransition": "none",
        "height": "auto",
        "maxHeight": "300px",
        "msTransition": "none",
        "overflow": undefined,
        "overflowY": "auto",
        "transition": "none",
      }
    }
  >
    <div>
      <styled.div
        key="123-title"
        onClick={[Function]}
        tabIndex="0"
        title="item1"
      >
        <styled.label>
          item1
        </styled.label>
      </styled.div>
      <styled.div
        key="321-title"
        onClick={[Function]}
        tabIndex="0"
        title="item2"
      >
        <styled.label>
          item2
        </styled.label>
      </styled.div>
    </div>
  </div>
</div>
`;
